### PR TITLE
feat: Add volunteers to user management table

### PR DESCRIPTION
**Pull request by AmandaMisjuwar**
## Implementation description. How did you make this change?
* Changed user management table data to be a merged list of donors and volunteers
* Updated table accordingly based on designs
* Called API functions to approve volunteers and delete users
* Added filter by role

## Steps to test
1. Make sure database has both users and donors
2. Log in as an admin 
3. Go to User Management page
4. Make sure table shows all users and their info is correct
5. Volunteers with 'Approved' status should have nothing under the 'Approvals' column.
6. Volunteers with 'Pending' or 'Rejected' status should show a 'Approve' button
7. Volunteers should have blank company column
8. Donors should have blank 'Approvals' column
9. Approving a volunteer changes status accordingly
10. Deleting a user deletes the correct user.
11. Searching and filtering work as needed

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages follow conventional commits and are descriptive. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s) 
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR
- [ ] The appropriate tests if necessary have been written

### CODE REVIEW

File: frontend/src/components/pages/Account.tsx [hunk 1]
```diff
@@ -75,7 +74,11 @@ const Account = (): JSX.Element => {
         );
       }
       setDonor(donorResponse);
-   setBusinessName(authenticatedUser?.role !== Role.VOLUNTEER ? donorResponse.businessName: "");
+      setBusinessName(
+        authenticatedUser?.role !== Role.VOLUNTEER
+          ? donorResponse.businessName
+          : "",
+      );
```
Comment (AmandaMisjuwar) [inline frontend/src/components/pages/Account.tsx:81]: Idk why this change is in here :'(((

File: frontend/src/components/pages/UserManagement/index.tsx [hunk 2]
```diff
@@ -15,68 +19,203 @@ import {
...
+    );
+    setDataChanged(dataChanged === 0 ? 1 : 0);
+    console.log(dataChanged);
+  };
+
+  // Deletes selected user
+  const handleDeleteUser = (user: UserMgmtTableRecord) => {
+    const updatedUser = UserAPIClient.deleteUserById(user.userId);
+    setDataChanged(dataChanged === 0 ? 1 : 0);
+  };
```
Comment (AmandaMisjuwar) [inline frontend/src/components/pages/UserManagement/index.tsx:140]: Calling `UserAPIClient.deleteUserById(user.userId)` for now, all of the delete stuff should be handled in this function @Shehryar21

File: frontend/src/APIClients/UserAPIClient.ts [hunk 3]
```diff
@@ -60,4 +60,20 @@ const updateUserById = async (
...
+  try {
+    await baseAPIClient.delete(`/users?userId=${id}`, {
+      headers: { Authorization: bearerToken },
+    });
+    return true;
+  } catch (error) {
+    return false;
+  }
+};
+
```
Comment (AmandaMisjuwar) [inline frontend/src/APIClients/UserAPIClient.ts:78]: this function will be overwritten

File: frontend/src/components/pages/UserManagement/index.tsx [hunk 4]
```diff
@@ -15,68 +19,203 @@ import {
...
+
+  // Sets volunteer status to Approved
+  const handleApprove = (user: UserMgmtTableRecord) => {
+    const newVolunteerData = { status: Status.APPROVED };
+    const updatedVolunteer = VolunteerAPIClient.updateVolunteerById(
+      user.id,
+      newVolunteerData,
+    );
+    setDataChanged(dataChanged === 0 ? 1 : 0);
+    console.log(dataChanged);
```
Comment (Shehryar21) [inline frontend/src/components/pages/UserManagement/index.tsx:133]: remove plz

File: frontend/src/components/pages/UserManagement/index.tsx (hunk 2 above)
Comment (Shehryar21) [reply frontend/src/components/pages/UserManagement/index.tsx:140]: Yeah thats all that needs to be done here.

File: frontend/src/APIClients/UserAPIClient.ts (hunk 3 above)
Comment (Shehryar21) [reply frontend/src/APIClients/UserAPIClient.ts:78]: this will remain the same, just the backend function corresponding to the route that this function calls is being changed accordingly. A pr is open for that (https://github.com/uwblueprint/community-fridge-kw/pull/184). We can do the end to end testing of delete in the pr that gets merged in after. It will make testing a bit easier in the other pr if we merge this pr first (cuz dont have to delete users using postman)

**Review by Shehryar21 — CHANGES_REQUESTED** [review]
One issue that i noticed functionality wise:
When I select volunteers in the account type dropdown, and then search, it searches on all accounts not just volunteers and vice versa. See screenshot below:
image

Approve worked fine for me and I could see both donors and volunteers in the table, nice work!!

File: frontend/src/App.tsx [hunk 5]
```diff
@@ -58,68 +58,70 @@ const App = (): React.ReactElement => {
...
-                exact
-                path={Routes.FORGET_PASSWORD}
-                component={ResetPassword}
-              />
-              <Route exact path="*" component={NotFound} />
-            </Switch>
-            <Footer />
-          </Router>
-        </AuthContext.Provider>
+          <AuthContext.Provider
```
Comment (hanlinc27) [inline frontend/src/App.tsx]: nothing's changed with this page? seems like `yarn run fix` is needed to clean up linting and it'll probably revert this diff

File: frontend/src/components/pages/UserManagement/index.tsx [hunk 6]
```diff
@@ -15,68 +19,202 @@ import {
...
+    {
+      value: "volunteers",
+      label: "Volunteers",
+    },
+    {
+      value: "donors",
+      label: "Donors",
+    },
+  ];
+
```
Comment (hanlinc27) [inline frontend/src/components/pages/UserManagement/index.tsx:51]: can we move this to `types` as well like the UserMgmtTableRecord?

File: frontend/src/components/pages/UserManagement/index.tsx [hunk 7]
```diff
@@ -15,68 +19,202 @@ import {
...
-        donor.firstName
-          .toLowerCase()
-          .concat(" ", donor.lastName.toLowerCase())
-          .includes(search.toLowerCase()) ||
-        donor.phoneNumber.toLowerCase().includes(search.toLowerCase()),
-    );
-  }, [search, donors]);
+  const filterOptions = [
+    {
+      value: "all",
```
Comment (hanlinc27) [inline frontend/src/components/pages/UserManagement/index.tsx:39]: small nit but I think we reference these values as just plain string a few times for us to maybe extract it into a enum? 
```suggestion
      value: "all",
```

File: frontend/src/components/pages/UserManagement/index.tsx [hunk 8]
```diff
@@ -15,68 +19,202 @@ import {
...
+        id: volunteer.id,
+        pointOfContact: `${volunteer.firstName} ${volunteer.lastName}`,
+        company: "",
+        email: volunteer.email,
+        phoneNumber: volunteer.phoneNumber,
+        accountType: volunteer.role,
+        approvalStatus: volunteer.status,
+      });
+    });
+
```
Comment (hanlinc27) [inline frontend/src/components/pages/UserManagement/index.tsx:68]: do you know if the volunteers that are pending should be taking precedence in the table?

File: frontend/src/components/pages/UserManagement/index.tsx [hunk 9]
```diff
@@ -15,68 +19,202 @@ import {
...
     getDonors();
-  }, []);
+    getVolunteers();
+  }, [dataChanged]);
+
+  // Update users whenever donors or volunteers have a change
+  React.useEffect(() => {
+    setUsers(buildTableUsers());
+  }, [donors, volunteers]);
```
Comment (hanlinc27) [inline frontend/src/components/pages/UserManagement/index.tsx:103]: volunteers should never change right?
```suggestion
  React.useEffect(() => {
    setUsers(buildTableUsers());
  }, [volunteers]);
```

File: frontend/src/components/pages/UserManagement/index.tsx [hunk 10]
```diff
@@ -15,68 +19,202 @@ import {
...
+        (user) => user.accountType === Role.VOLUNTEER,
+      );
+    else
+      filteredUsers = users.filter((user) => user.accountType === Role.DONOR);
+
+    if (!search) return filteredUsers;
+    return users.filter(
+      (user) =>
+        user.pointOfContact.toLowerCase().includes(search.toLowerCase()) ||
+        user.company.toLowerCase().includes(search.toLowerCase()) ||
```
Comment (hanlinc27) [inline frontend/src/components/pages/UserManagement/index.tsx:119]:
```suggestion
      (user) =>
        user.accountType.toLowerCase().includes(search.toLowerCase()) ||
        user.pointOfContact.toLowerCase().includes(search.toLowerCase()) ||
        user.company.toLowerCase().includes(search.toLowerCase()) ||
```
because:
No search/filter state:
image
"volunteer" entered in search: (showing 2 results instead of 3)
image

File: frontend/src/components/pages/UserManagement/index.tsx [hunk 11]
```diff
@@ -15,68 +19,202 @@ import {
...
+        user.email.toLowerCase().includes(search.toLowerCase()) ||
+        user.phoneNumber.toLowerCase().includes(search.toLowerCase()),
+    );
+  }, [search, selectedFilter, users]);
+
+  // Sets volunteer status to Approved
+  const handleApprove = (user: UserMgmtTableRecord) => {
+    const newVolunteerData = { status: Status.APPROVED };
+    const updatedVolunteer = VolunteerAPIClient.updateVolunteerById(
```
Comment (hanlinc27) [inline frontend/src/components/pages/UserManagement/index.tsx:128]: this isn't being used?

File: frontend/src/components/pages/UserManagement/index.tsx [hunk 12]
```diff
@@ -15,68 +19,202 @@ import {
...
+import { VolunteerResponse } from "../../../types/VolunteerTypes";
+import { UserMgmtTableRecord } from "./types";
 
 const UserManagementPage = (): JSX.Element => {
   const [donors, setDonors] = React.useState([] as DonorResponse[]);
+  const [volunteers, setVolunteers] = React.useState([] as VolunteerResponse[]);
+  const [users, setUsers] = React.useState([] as UserMgmtTableRecord[]); // both donors and volunteers
   const [search, setSearch] = React.useState("");
+  const [selectedFilter, setSelectedFilter] = React.useState("all");
+  const [dataChanged, setDataChanged] = React.useState<0 | 1>(0);
```
Comment (hanlinc27) [inline frontend/src/components/pages/UserManagement/index.tsx:35]: I don't think we should use this kind of state handling to watch for if there's changes to volunteer approval or to user deletion.
By running a `useEffect` and making a whole new backend call to refresh the table for these changes then building it all over again with `buildTableUsers` is expensive.

We can delete all of:
```typescript
  React.useEffect(() => {
    const getDonors = async () => {
      const res = await DonorAPIClient.getAllDonors();
      setDonors(res);
    };

    const getVolunteers = async () => {
      const res = await VolunteerAPIClient.getAllVolunteers();
      setVolunteers(res);
    };

    getDonors();
    getVolunteers();
  }, [dataChanged]);
```
and instead for **`handleDeleteUser`** and **`handleApprove`** we filter in-place the the `users`state.
i.e. 
```typescript
   const handleDeleteUser = async (user: UserMgmtTableRecord) => {
    const deleteUserResponse = await UserAPIClient.deleteUserById(user.userId);
    if (deleteUserResponse) {
    users.filter((u)=>u.id===user.id)
}
  };

  const handleApprove =  async (user: UserMgmtTableRecord) => {
    const newVolunteerData = { status: Status.APPROVED };
    const updatedVolunteerResponse = await VolunteerAPIClient.updateVolunteerById(
      user.id,
      newVolunteerData,
    );
    users.find(u => u.id === updatedVolunteerResponse.userId).status=Status.APPROVED;

  };
```

**Review by hanlinc27 — CHANGES_REQUESTED** [review]
first pass, let me know if any of the comments are confusing!

File: frontend/src/components/pages/UserManagement/index.tsx (hunk 12 above)
Comment (Shehryar21) [reply frontend/src/components/pages/UserManagement/index.tsx:35]: for the approach that Hanlin suggested for `handleApprove`, just one small addition to check for error so that we are not showing any wrong information on frontend.

File: frontend/src/components/pages/UserManagement/index.tsx (hunk 8 above)
Comment (AmandaMisjuwar) [reply frontend/src/components/pages/UserManagement/index.tsx:68]: @hanlinc27 getting confirmation on this 

(originally we wanted sorted by date created, and approval status doesn't matter (so no to pending taking precedence), but currently i have that volunteers are on top since the donors are appended to this list

Comment (AmandaMisjuwar) [reply frontend/src/components/pages/UserManagement/index.tsx:68]: Anya says this is fine

File: frontend/src/components/pages/UserManagement/index.tsx [hunk 13]
```diff
@@ -15,68 +19,206 @@ import {
...
+    console.log("ON APPROVE CLICK");
+    const newVolunteerData = { status: Status.APPROVED };
+    const updatedVolunteerResponse: VolunteerDTO = await VolunteerAPIClient.updateVolunteerByUserId(
+      user.userId,
+      newVolunteerData,
+    );
+    console.log("AFTER BE CALL");
+    console.log("updatedResponse", updatedVolunteerResponse);
+    if (updatedVolunteerResponse) {
+      const i = users.findIndex((u) => u.id === user.userId);
```
Comment (AmandaMisjuwar) [inline frontend/src/components/pages/UserManagement/index.tsx:128]: 04.01 - having an issue with approving volunteers. We want the 'Approve' button to be gone once an admin has clicked it and changed the selected volunteer's approval status. On clicking, we do hit `updateVolunteerByUserId` and it correctly updates the database. But it doesn't return a response (returns nothing), which is causing problems here since `updatedVolunteerResponse` is not there. But we updated `VolunteerAPIClient` and `VolunteerService.updateVolunteerByUserId` to return stuff

Comment (Shehryar21) [reply frontend/src/components/pages/UserManagement/index.tsx]: the new commit should fix this issue

File: backend/typescript/services/implementations/donorService.ts [hunk 14]
```diff
@@ -173,6 +166,13 @@ class DonorService implements IDonorService {
         );
         throw error;
       }
+      const numDestroyed: number = await Donor.destroy({
```
Comment (Shehryar21) [inline backend/typescript/services/implementations/donorService.ts:169]: we need delete schedules before we delete donors cuz if we dont, we will get foreign key constraint error. Same for volunteers, which is why I added this change in both of the delete functions.

**Review by Shehryar21 — APPROVED** [review]
Sometimes when I am deleting, I get unauthorized error even though I am signed in as admin. After refreshing the page and trying again, it works fine. That is not an issue of this PR though. Not sure why this is happening or if its only happening to me but we can look at it later (i have created a ticket for it).

**Review by Shehryar21 — APPROVED** [review]

**Review by hanlinc27 — APPROVED** [review]
